src/charm.py

```diff
@@ -98,68 +99,80 @@ def _configure_pod(self):
+                "Error fetching image information"
+            )
             return {}
 
         config = self.model.config
         self.unit.status = WaitingStatus("Assembling pod spec")
+
         pod_spec = {
-            'version': 3,
-            'containers': [{
-                'name': self.app.name,
-                'imageDetails': image_info,
-                'ports': [{
-                    'containerPort': config['port'],
-                    'protocol': 'TCP'
-                }],
-                'envConfig': {
-                    'MYSQL_ROOT_PASSWORD': config['root-password'],
-                },
-            }]
+            "version": 3,
+            "containers": [
+                {
+                    "name": self.app.name,
+                    "imageDetails": image_info,
+                    "ports": [
+                        {"containerPort": config["port"], "protocol": "TCP"}
+                    ],
+                    "envConfig": self.env_config,
+                }
+            ],
         }
 
         return pod_spec
 
-    def _mysql_is_ready(self):
+    def _mysql_is_ready(self) -> bool:
         """Check that every unit has mysql running.
 
         Until we have a good event-driven way of using the Kubernetes
         readiness probe, we will attempt
         """
-        for hostname in self.hostnames:
-            try:
-                cnx = self._get_sql_connection_for_host(hostname)
-                logger.warning("MySQL service is ready for {}.".format(hostname))
-            except mysql.connector.Error:
-                # TODO: Improve exceptions handling
-                logger.warning("MySQL service is not ready for {}.".format(hostname))
-                return False
-            else:
-                cnx.close()
+        try:
+            cnx = self._get_sql_connection_for_host()
+            logger.info("MySQL service is ready in %s.", self.hostname)
+        except mysql.connector.Error:
+            # TODO: Improve exceptions handling
+            logger.info("MySQL service is not ready in %s", self.hostname)
+            return False
+        else:
+            cnx.close()
 
         return True
 
-    def _get_sql_connection_for_host(self, hostname):
+    def _get_sql_connection_for_host(self):
+        """Helper for the _mysql_is_ready() method"""
         config = {
-            'user': 'root',
-            'password': self.model.config['root-password'],
-            'host': hostname,
-            'port': self.model.config['port']
+            "user": "root",
+            "password": self._stored.MYSQL_ROOT_PASSWORD,
+            "host": self.hostname,
+            "port": self.model.config["port"],
         }
         return mysql.connector.connect(**config)
 
-    def _get_unit_hostname(self, _id: int) -> str:
+    def _get_hostname_from_unit(self, _id: int) -> str:
         """Construct a DNS name for a MySQL unit."""
         return "{0}-{1}.{0}-endpoints".format(self.model.app.name, _id)
```

Comment:
I don't think this will work. The unit number does not always start at 0 but the hostnames do. For example, in my currently deployed charm, the unit name is cassandra/44 while the hostname is cassandra-0.cassandra-endpoints.lma.svc.cluster.local.

Reply:
Once we are able to run code on in the application pod, you will be able to get this by looking in /etc/hosts

Reply:
> I don't think this will work. The unit number does not always start at 0 but the hostnames do. For example, in my currently deployed charm, the unit name is cassandra/44 while the hostname is cassandra-0.cassandra-endpoints.lma.svc.cluster.local.

Oh! I thought that **always** the unit number was the same as the number in the hostname.
Maybe I did not realise that because in every test I destroyed the model, and add it again

Reply:
I think I can get the internal address with this:

```python
    @property
    def bind_address(self) -> str:
        """The ingress-address of the MySQL server"""
        return str(self.model.get_binding(PEER).network.bind_address)
```

It seems it's working...

Reply:
Fixed